libs/model/src/middleware/tiered.ts

```diff
@@ -0,0 +1,109 @@
+import {
+  CacheNamespaces,
+  Context,
+  InvalidActor,
+  cache,
+} from '@hicommonwealth/core';
+import moment from 'moment';
+import { Op } from 'sequelize';
+import { ZodSchema } from 'zod';
+import { models } from '../database';
+
+const tierLimitsPerHour = [
+  // tier 0 (new unverified addresses with no sso, balance, etc - bad aura)
+  { create: 0, upvote: 0, ai: { images: 0, text: 0 } },
+  // tier 1 (new verified address)
+  { create: 1, upvote: 5, ai: { images: 2, text: 5 } },
+  // tier 2 (1 week old and verified)
+  { create: 2, upvote: 10, ai: { images: 4, text: 10 } },
+  // tier 3 (email verified/sso, or wallet with balance (TODO: cache balances))
+  { create: 5, upvote: 25, ai: { images: 10, text: 50 } },
+];
+
+function builtKey(user_id: number, counter: 'creates' | 'upvotes') {
+  return `${user_id}-${counter}-${new Date().toISOString().substring(0, 13)}`;
+}
+
+async function getUserCount(user_id: number, counter: 'creates' | 'upvotes') {
+  const cacheKey = builtKey(user_id, counter);
+  const count = await cache().getKey(CacheNamespaces.TieredCounter, cacheKey);
+  return +(count ?? '0');
+}
+
+export async function incrementUserCount(
+  user_id: number,
+  counter: 'creates' | 'upvotes',
+) {
+  const cacheKey = builtKey(user_id, counter);
+  try {
+    const requestCount = await cache().incrementKey(
+      CacheNamespaces.TieredCounter,
+      cacheKey,
+    );
+    if (requestCount === 1) {
+      // is first request in window, set expiration
+      await cache().setKeyTTL(CacheNamespaces.TieredCounter, cacheKey, 60);
```

Comment:
If this request fails then the key-value pair will persist forever in Redis and we will need to have a manual job to clean these up. What we can do to avoid this is use `INCR` alongside `EXPIRE` in a transaction (multi) and provide the `NX` option to the `EXPIRE` command which means 'set expiry only when the key has no expiry'. This way any key created is guaranteed to expire and the memory can be reclaimed.

Reply:
yeah, I tried that but didn't know we had this feature... where can I find an example to replicate?